redis/src/sentinel.rs

```diff
@@ -0,0 +1,768 @@
+/// servers (masters and replicas) belonging to it.
+#[derive(Clone, Default)]
+pub struct SentinelNodeConnectionInfo {
+    /// The TLS mode of the connection, or None if we do not want to connect using TLS
+    /// (just a plain TCP connection).
+    pub tls_mode: Option<TlsMode>,
+
+    /// The Redis specific/connection independent information to be used.
+    pub redis_connection_info: Option<RedisConnectionInfo>,
+}
+
+impl SentinelNodeConnectionInfo {
+    fn create_connection_info(&self, ip: String, port: u16) -> ConnectionInfo {
+        let addr = match self.tls_mode {
+            None => crate::ConnectionAddr::Tcp(ip, port),
+            Some(TlsMode::Secure) => crate::ConnectionAddr::TcpTls {
+                host: ip,
+                port,
+                insecure: false,
+            },
+            Some(TlsMode::Insecure) => crate::ConnectionAddr::TcpTls {
+                host: ip,
+                port,
+                insecure: true,
+            },
+        };
+
+        ConnectionInfo {
+            addr,
+            redis: self.redis_connection_info.clone().unwrap_or_default(),
+        }
+    }
+}
+
+impl Default for &SentinelNodeConnectionInfo {
+    fn default() -> Self {
+        static DEFAULT_VALUE: SentinelNodeConnectionInfo = SentinelNodeConnectionInfo {
+            tls_mode: None,
+            redis_connection_info: None,
+        };
+        &DEFAULT_VALUE
+    }
+}
+
+fn sentinel_masters_cmd() -> crate::Cmd {
+    let mut cmd = crate::cmd("SENTINEL");
+    cmd.arg("MASTERS");
+    cmd
+}
+
+fn sentinel_replicas_cmd(master_name: &str) -> crate::Cmd {
+    let mut cmd = crate::cmd("SENTINEL");
+    cmd.arg("SLAVES"); // For compatibility with older redis versions
+    cmd.arg(master_name);
+    cmd
+}
+
+fn is_master_valid(master_info: &HashMap<String, String>, service_name: &str) -> bool {
+    master_info.get("name").map(|s| s.as_str()) == Some(service_name)
+        && master_info.contains_key("ip")
+        && master_info.contains_key("port")
+        && master_info.get("flags").map_or(false, |flags| {
+            flags.contains("master") && !flags.contains("s_down") && !flags.contains("o_down")
+        })
+        && master_info["port"].parse::<u16>().is_ok()
+}
+
+fn is_replica_valid(replica_info: &HashMap<String, String>) -> bool {
+    replica_info.contains_key("ip")
+        && replica_info.contains_key("port")
+        && replica_info.get("flags").map_or(false, |flags| {
+            !flags.contains("s_down") && !flags.contains("o_down")
+        })
+        && replica_info["port"].parse::<u16>().is_ok()
+}
+
+/// Generates a random value in the 0..max range. If max is zero, the returned value is
+/// also zero.
+fn random_replica_index(max: usize) -> usize {
+    if max == 0 {
+        // This is just to avoid panicking in the gen_range function, this value does
+        // not have any special meaning, and is not expected to be used.
+        0
```

Comment:
use Option in order to better convey these semantics.

Reply:
What about NonZeroUsize? After your comment I thought about it, just pushed a commit with it. Not sure if there is any downside to using it though.